pkg/apis/workflow/v1alpha1/cluster_workflow_template_types.go

```diff
@@ -0,0 +1,62 @@
+package v1alpha1
+
+import (
+	"strings"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ClusterWorkflowTemplate is the definition of a workflow template resource in cluster scope
+// +genclient
+// +genclient:noStatus
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterWorkflowTemplate struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Spec              WorkflowTemplateSpec `json:"spec" protobuf:"bytes,2,opt,name=spec"`
+}
+
+type ClusterWorkflowTemplates []ClusterWorkflowTemplate
+
+func (w ClusterWorkflowTemplates) Len() int {
+	return len(w)
+}
+
+func (w ClusterWorkflowTemplates) Less(i, j int) bool {
+	return strings.Compare(w[j].ObjectMeta.Name, w[i].ObjectMeta.Name) > 0
+}
+
+func (w ClusterWorkflowTemplates) Swap(i, j int) {
+	w[i], w[j] = w[j], w[i]
+}
+
+// ClusterWorkflowTemplateList is list of ClusterWorkflowTemplate resources
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+type ClusterWorkflowTemplateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata" protobuf:"bytes,1,opt,name=metadata"`
+	Items           ClusterWorkflowTemplates `json:"items" protobuf:"bytes,2,rep,name=items"`
+}
+
+var _ TemplateGetter = &ClusterWorkflowTemplate{}
+
+// GetTemplateByName retrieves a defined template by its name
+func (cwftmpl *ClusterWorkflowTemplate) GetTemplateByName(name string) *Template {
+	for _, t := range cwftmpl.Spec.Templates {
+		if t.Name == name {
+			return &t
+		}
+	}
+	return nil
+}
+
+// GetTemplateScope returns the template scope of workflow template.
+func (cwftmpl *ClusterWorkflowTemplate) GetTemplateScope() string {
+	return "cluster/" + cwftmpl.Name
```

Comment:
This is a bit pedantic, but we use different grammar with `cluster/<template_name>` and `namespaced/<template_name>`.

In my opinion it should be either:

- `clustered/<template_name>` and `namespaced/<template_name>`; or
- `cluster/<template_name>` and `namespace/<template_name>`

Reply:
I am just following Kubernetes conventions 

```
 # either Namespaced or Cluster
  scope: Namespaced
```

Reply:
```
// ResourceScope is an enum defining the different scopes available to a custom resource
type ResourceScope string

const (
	ClusterScoped   ResourceScope = "Cluster"
	NamespaceScoped ResourceScope = "Namespaced"
)
```

Reply:
@jessesuen I would like to get your feedback on this before merging this PR.

Reply:
Where is this `GetTemplateScope` method even being used?

Reply:
Also, based on the name of this method, I would expect it to either return `Cluster` or `Namespaced`, but I see it's the scope + name of the resource, which they could already get by `metadata.name`

Can I learn the intended use of this method?

Reply:
it is used in multiple where the controller tries to load the stored template.
https://github.com/argoproj/argo/blob/184c3653085bc8821bdcd65f5476fbe24f24b00e/workflow/templateresolution/context.go#L152
https://github.com/argoproj/argo/blob/fb74ba1ce27b96473411c2c5cfe9a86972af589e/workflow/controller/steps.go#L199

Reply:
I see, I thought it was being introduced in this change, but could not find the usage in this PR. In any case, I think following kubernetes scope naming of `cluster` and `namespaced` is the right choice.

That said, I do believe it would be better to make `GetTemplateScope()` return a ResourceScope kind, change `GetStoredTemplate()` to accept both a scope and name, and formulate the map key based on concatenating the two.

Reply:
I agreed with you.  But `TemplateScope` word  is used across workflowtemplate implantation and workflow spec. Instead of changing one place. I would recommend adding a new function that will return templateBaseID (scope + name) and rename variable, arguments (except spec change). I would like to do it in separate PR, so, it will be easier to review.